ste/sender-azureFile.go

```diff
@@ -34,14 +34,27 @@ import (
 	"github.com/Azure/azure-storage-azcopy/common"
 )
 
+type URLHolder interface {
+	URL() url.URL
+	String() string
+}
+
+// TODO: review the decision to have this implement both interfaces
+// azureFileSenderBase implements both IFolderSender and (most of) IFileSender.
+// Why implement both interfaces in the one type, even though they are largely unrelated? Because it
+// makes functions like newAzureFilesUploader easier to reason about, since they always return the same type.
+// It may also make it easier to describe what's needed when supporting an new backend - e.g. "to send to a new back end
+// you need a sender that implements IFileSender and, if the back end is folder aware, it should also implement IFolderSender"
+// (The alternative would be to have the likes of newAzureFilesUploader call sip.EntityType and return a different type
+// if the entity type is folder).
 type azureFileSenderBase struct {
-	jptm      IJobPartTransferMgr
-	fileURL   azfile.FileURL
-	chunkSize uint32
-	numChunks uint32
-	pipeline  pipeline.Pipeline
-	pacer     pacer
-	ctx       context.Context
+	jptm         IJobPartTransferMgr
+	fileOrDirURL URLHolder
```

Comment:
>fileOrDirURL URLHolder [](start = 1, length = 22)

Some work may need to be done on the user delegation PR to fit compatibility with this.

Reply:
OK.  How about we get that PR done and merged before this one?  Would that be sensible?

Reply:
That sounds good to me. So that'd make the order of these SDDLs -> Folders -> User Delegation.

---
In reply to: [369332636](https://github.com/Azure/azure-storage-azcopy/pull/824#discussion_r369332636) [](ancestors = 369332636)